ondemand_agentmanager.go

```diff
@@ -72,60 +77,94 @@ func (m *OnDemandAgentManager) Reconfigure(opts OnDemandAgentManagerReconfigureO
 
 // GetClusterAgent returns the Agent which is not bound to any bucket.
 func (m *OnDemandAgentManager) GetClusterAgent() (*Agent, error) {
-	m.lock.Lock()
-	defer m.lock.Unlock()
-
+	m.stateLock.Lock()
 	if m.closed {
+		m.stateLock.Unlock()
 		return nil, errors.New("agent manager closed")
 	}
+	m.stateLock.Unlock()
 
 	return m.clusterAgent, nil
 }
 
 // GetBucketAgent returns the Agent bound to a specific bucket, creating a new Agent as required.
 func (m *OnDemandAgentManager) GetBucketAgent(ctx context.Context, bucketName string) (*Agent, error) {
-	m.lock.Lock()
-	defer m.lock.Unlock()
-
+	m.stateLock.Lock()
 	if m.closed {
+		m.stateLock.Unlock()
 		return nil, errors.New("agent manager closed")
 	}
+	m.stateLock.Unlock()
 
-	if m.bucketAgents == nil {
-		m.bucketAgents = make(map[string]*Agent)
-	} else {
-		bucketAgent, ok := m.bucketAgents[bucketName]
+	fastMap := m.fastMap.Load()
+	if fastMap != nil {
+		agent, ok := (*fastMap)[bucketName]
 		if ok {
-			return bucketAgent, nil
+			return agent, nil
 		}
 	}
 
-	bucketAgent, err := m.makeAgentLocked(ctx, bucketName)
+	agent, err := m.getBucketAgentSlow(ctx, bucketName)
+	if err != nil {
+		return nil, err
+	}
+
+	return agent, nil
+}
+
+func (m *OnDemandAgentManager) getBucketAgentSlow(ctx context.Context, bucketName string) (*Agent, error) {
+	m.slowLock.Lock()
+	defer m.slowLock.Unlock()
+	if agent, ok := m.slowMap[bucketName]; ok {
+		return agent, nil
+	}
+
+	agent, err := m.makeAgent(ctx, bucketName)
```

Comment:
We should file a bug to someday improve this even more and allow multiple new agents to be built simultaneously.